File: GaniOSApp/GanClient/GanClient/Gan/Service/CallbackService.swift
//
//  CallbackService.swift
//  GanClient
//
//  Created by xattacker on 2020/11/25.
//

import Foundation


internal protocol CallbackServiceDelegate: class
{
    func onLoggedOut(account: String)
    func onMessageReceived(message: MessageData)
}


internal final class CallbackService: ServiceFoundation
{
    private let delegate: CallbackServiceDelegate
    private var task: CallbackReceivingTask?
    
    internal init(agent: GanAgent, delegate: CallbackServiceDelegate)
    {
        self.delegate = delegate
        
        super.init(agent: agent)
    }
    
    internal func handleConnection(_ connection: SocketConnection)
    {
        self.task = CallbackReceivingTask(connection: connection, callbackDelegate: self.delegate)
        self.task?.start()
    }
    
    internal func disconnect()
    {
        self.task?.close()
        self.task = nil
    }
    
    deinit
    {
        self.disconnect()
    }
}


internal final class CallbackReceivingTask: ImpThread
{
    private var connection: SocketConnection?
    private weak var callbackDelegate: CallbackServiceDelegate?

    init(connection: SocketConnection, callbackDelegate: CallbackServiceDelegate)
    {
        self.connection = connection
        self.callbackDelegate = callbackDelegate
    }
    
    override func run() throws
    {
        repeat
        {
            self.sleep(0.3)
            
            guard let connection = self.connection else
            {
                continue
            }
            
            let valid = PackChecker.isValidPack(connection)
            if valid.valid && valid.length > 0
            {
                guard let data = connection.read(valid.length, timeout: 3) else
                {
                    print("read timeout, available: \(connection.available)")
                    continue
                }
                
                
                let buffer = BinaryBuffer(bytes: data, length: UInt(data.count))
                let response = ResponsePack()
                if response.fromBinaryReadable(buffer)
                {
                    switch FunctionType.init(rawValue: response.id)
                    {
                        case .receive_sms:
                            guard let json = response.responseString,
                                  let msg = MessageData(JSONString: json) else
                            {
                                break
                            }
                            
                            // send Ack response, but server side could not receive it ??!!
                            let ack = MessageAck()
                            ack.id = msg.id
                            
                            guard let ack_data = ack.toJSONString()?.data(using: String.Encoding.utf8) else
                            {
                                break
                            }
                            
                            let buffer_ack = BinaryBuffer()
                            PackChecker.packData(ack_data, container: buffer_ack)
                            
                            switch connection.send(data: buffer_ack.data)
                            {
                                case .success(()):
//                                    Thread.sleep(forTimeInterval: 0.5)
//
//                                    let valid = PackChecker.isValidPack(connection)
//                                    guard valid.valid && valid.length > 0 else
//                                    {
//                                        continue
//                                    }
//
//                                    guard let data2 = connection.read(valid.length, timeout: 5) else
//                                    {
//                                        continue
//                                    }
//
//                                    let buffer2 = BinaryBuffer(bytes: data2, length: UInt(data2.count))
//                                    let response2 = ResponsePack()
//                                    if response2.fromBinaryReadable(buffer2), response2.result
//                                    {
//                                        self.callbackDelegate?.onMessageReceived(message: msg)
//                                    }
                                    
                                    self.callbackDelegate?.onMessageReceived(message: msg)
                                    Thread.sleep(forTimeInterval: 0.2)
                                    
                                case .failure(let error):
                                    print(error)
                                    continue
                            }
                            break
                            
                        case .logout: // force logout
                            if let account = response.responseString
                            {
                                self.callbackDelegate?.onLoggedOut(account: account)
                            }
                            
                            self.connection?.close()
                            self.connection = nil
                            break

                        default:
                            print("unknown response id: \(response.id)")
                            break
                    }
                }
            }
            
        } while self.connection != nil && !self.isTerminated
    }
    
    deinit
    {
        self.connection?.close()
        self.connection = nil
        
        self.delegate = nil
    }
}

File: GanAndroidApp/GanClient/src/main/java/com/xattacker/gan/service/CallbackService.kt
package com.xattacker.gan.service

import android.util.Log
import com.xattacker.binary.InputBinaryBuffer
import com.xattacker.gan.GanAgent
import com.xattacker.gan.data.*
import com.xattacker.json.JsonUtility
import java.lang.ref.WeakReference
import java.net.Socket

internal interface CallbackServiceListener
{
    fun onLoggedOut(account: String)
    fun onMessageReceived(message: MessageData)
}


internal class CallbackService internal constructor(agent: GanAgent, listener: CallbackServiceListener) : ServiceFoundation(agent), Runnable
{
    private var socket: Socket? = null
    private var thread: Thread? = null
    private var listener: WeakReference<CallbackServiceListener>? = null

    private val ENABLE_MSG_ACK = false

    init
    {
        this.listener = WeakReference(listener)
    }

    fun handleConnection(socket: Socket)
    {
        this.socket = socket
        this.socket?.keepAlive = true
        this.socket?.soTimeout = 0
        this.socket?.oobInline = true

        thread = Thread(this)
        thread?.start()
    }

    fun close()
    {
        thread = null

        socket?.close()
        socket = null
    }

    override fun run()
    {
        try
        {
            val out = socket?.getOutputStream()
            val ins = socket?.getInputStream()
            if (out == null || ins == null)
            {
                return
            }


            while (socket != null)
            {
                Thread.sleep(500)

                val valid = PackChecker.isValidPack(ins, 50)
                if (valid.valid && valid.length > 0)
                {
                    wait(ins, valid.length, 50, socket?.receiveBufferSize ?: 0)


                    val binary = InputBinaryBuffer(ins)
                    val response = ResponsePack()
                    response.fromBinary(binary)

                    if (response.result)
                    {
                        when (FunctionType.parse(response.id))
                        {
                            FunctionType.RECEIVE_SMS ->
                            {
                                response.response?.let {
                                    val json = String(it, Charsets.UTF_8)
                                    val msg = JsonUtility.fromJson(json, MessageData::class.java)
                                    listener?.get()?.onMessageReceived(msg)

                                    if (ENABLE_MSG_ACK)
                                    {
                                        val ack = MessageAck()
                                        ack.id = msg.id

                                        PackChecker.pack(ack.toJson().toByteArray(), out)
                                        Thread.sleep(200)
                                    }
                                }
                            }

                            FunctionType.LOGOUT ->
                            {
                                close()
                                listener?.get()?.onLoggedOut(this.agent.account ?: "")
                            }

                            else -> android.util.Log.d("aaa", "unknown callback response id: " + response.id)
                        }
                    }
                }
            }
        }
        catch (ex: Exception)
        {
            Log.i("aaa", ex.toString())
        }
    }
}